Guess the CSV character encoding if none is specified in Content-Type

mhl opened the pull request:
daff seems to error when trying to compare strings with encodings UTF-8
and ASCII-8BIT; this typically arises when the CSV files for comparison
is returned with a Content-Type header that omits the charset directive,
so it uses the 'just a sequence of bytes' encoding ASCII-8BIT. This
commit changes behaviour in that case so that the character set is
guessed using the charlock_holmes gem and the guessed encoding is used.

tmtmtmtm commented on lib/compare_with_wikidata.rb:
```diff
@@ -107,8 +108,14 @@ def expanded_wikitext(page_title)
+      return s unless s.encoding == Encoding::ASCII_8BIT
+      s.dup.force_encoding(s.detect_encoding[:ruby_encoding])
+    end
+
```
I'm not 100% convinced I'm not missing something here, but I think you can replace this with charlock holmes' `.detect_encoding!` method. 

If you're avoiding that because that's only available in ruby 1.9+, I think it's fine to make that a pre-requisite here too. If there's something subtler going on where this isn't actually equivalent, then we should have a test to catch that case, as they currently pass with that change.

tmtmtmtm commented on test/compare_with_wikidata_test.rb:
```diff
@@ -1,7 +1,28 @@
+    mediawiki_site: 'www.wikidata.org',
+    page_title: 'FooBar',
+  ) }
+
```
Having to add a new `subject` to an existing test is a bit of a red flag to me. If we want to add a test like this I think I'd add it either to a new test file entirely, or at least to a new `describe` block.

*However*, I'm also not entirely convinced that we do want tests like this. They're very implementation dependent (e.g. if we were to replace the `CSV.parse(fix_encoding(csv_text))` with `CSV.parse(csv_text.detect_encoding!)`, these tests would fail), and don't really exercise the actual problem we were facing (i.e. a comparison blowing up). 

I'd be a lot more comfortable if we added a test that actually takes two datasets with different encoding, and showed the comparison succeeding rather than exploding.

mhl replied:
As background, this wasn't how I really wanted to do this fix: it would be nice to do the encoding fixing in `daff_diff`, so that however that function is used in future can benefit from the change. And it's much easier to write specific tests for whether the comparison works or not. However, by that point any strings are spread across different cells - you could do something like:

```
    daff_table.map do |row|
      row.map do |cell|
        next cell unless cell.instance_of? String
        next cell unless cell.encoding == Encoding::ASCII_8BIT
        cell.dup.detect_encoding!
      end
    end
```

(full example [here](https://github.com/everypolitician/compare_with_wikidata/compare/fix-unknown-character-encoding2?expand=1)) but when you're considering a cell at a time, the character guessing of charlock_holmes works surprisingly badly (e.g. detecting UTF-8 as ISO-8859-1 in cases you're have thought were obvious); really you want to do the character detection on all the strings in the dataset to help its guesses. I guess you could still do it within `daff_diff`, by extracting all the strings in a dataset, concatenating them and guessing the encoding from that, but that seems pretty hacky.

However, it sounds as if you're OK with the implementation in this PR, and just asking for the tests to be more helpful. This wasn't particularly obvious because of how `CompareWithWikidata.run! ` is written, but I've added a refactoring commit first it a bit to make that possible. Also, I realised that there were lots of existing tests for `csv_from_url` so I've added a couple there too.

mhl commented:
@tmtmtmtm In fact, the higher level test I added (that of DiffOutputGenerator.comparison) wasn't failing for the right reason - it didn't actually trigger the "incompatible character encodings" because that was arising when the output was trying to format a change in a cell with `->`; unfortunately the original version of that test instead generated a diff which with `---` and `+++` output instead. I've fixed this now by making sure there is an ID in common between the two sides, so without the fix it fails for the right reason.

> If you agree with my comments there, but don't have time to add such a test yourself soon, I'm happy to add those myself to get this live quicker.

BTW, I'm very happy for you to rewrite these tests as you like - my curiousity about this bug is satisfied :)
